shell/platform/darwin/macos/framework/Source/FlutterMenuPlugin.mm

```diff
@@ -0,0 +1,418 @@
+      {PlatformProvidedMenu::showAllApplications, @selector(unhideAllApplications:)},
+      {PlatformProvidedMenu::startSpeaking, @selector(startSpeaking:)},
+      {PlatformProvidedMenu::stopSpeaking, @selector(stopSpeaking:)},
+      {PlatformProvidedMenu::toggleFullScreen, @selector(toggleFullScreen:)},
+      {PlatformProvidedMenu::minimizeWindow, @selector(performMiniaturize:)},
+      {PlatformProvidedMenu::zoomWindow, @selector(performZoom:)},
+      {PlatformProvidedMenu::arrangeWindowsInFront, @selector(arrangeInFront:)},
+  };
+}
+
+/*
+ * Returns the NSEventModifierFlags of |modifiers|, a value from
+ * kShortcutKeyModifiers.
+ */
+static NSEventModifierFlags KeyEquivalentModifierMaskForModifiers(NSNumber* modifiers) {
+  int flutterModifierFlags = modifiers.intValue;
+  NSEventModifierFlags flags = 0;
+  if (flutterModifierFlags & kFlutterShortcutModifierMeta) {
+    flags |= NSEventModifierFlagCommand;
+  }
+  if (flutterModifierFlags & kFlutterShortcutModifierShift) {
+    flags |= NSEventModifierFlagShift;
+  }
+  if (flutterModifierFlags & kFlutterShortcutModifierAlt) {
+    flags |= NSEventModifierFlagOption;
+  }
+  if (flutterModifierFlags & kFlutterShortcutModifierControl) {
+    flags |= NSEventModifierFlagControl;
+  }
+  // There are also modifier flags for things like the function (Fn) key, but
+  // the framework doesn't support those.
+  return flags;
+}
+
+/*
+ * An NSMenuDelegate used to listen for changes in the menu: when it opens and
+ * closes.
+ */
+@interface FlutterMenuDelegate : NSObject <NSMenuDelegate>
+/*
+ * When this delegate receives notification that the menu opened or closed, it
+ * will send a message on the given channel to that effect for the menu item
+ * with the given id (the ID comes from the data supplied by the framework to
+ * FlutterMenuPlugin.setMenu).
+ */
+- (instancetype)initWithIdentifier:(int64_t)identifier channel:(FlutterMethodChannel*)channel;
+@end
+
+@implementation FlutterMenuDelegate {
+  FlutterMethodChannel* _channel;
+  int64_t _identifier;
+}
+
+- (instancetype)initWithIdentifier:(int64_t)identifier channel:(FlutterMethodChannel*)channel {
+  self = [super init];
+  if (self) {
+    _identifier = identifier;
+    _channel = channel;
+  }
+  return self;
+}
+
+- (void)menuWillOpen:(NSMenu*)menu {
+  [_channel invokeMethod:kMenuItemOpenedMethod arguments:@(_identifier)];
+}
+
+- (void)menuDidClose:(NSMenu*)menu {
+  [_channel invokeMethod:kMenuItemClosedMethod arguments:@(_identifier)];
+}
+@end
+
+@interface FlutterMenuPlugin ()
+// Initialize the plugin with the given method channel.
+- (instancetype)initWithChannel:(FlutterMethodChannel*)channel;
+
+// Iterates through the given menu hierarchy, and replaces "APP_NAME"
+// with the localized running application name.
+- (void)replaceAppName:(NSArray<NSMenuItem*>*)items;
+
+// Look up the menu item with the given selector and tag in the list of provided
+// menus and return it.
```

Comment:
It would be useful to describe here how we're using `NSMenuItem`'s `tag` property.

Reply:
Actually, it is never called with anything other than zero anymore (there was a time...), so I just got rid of that parameter.